docs/source/user/extensions.rst

```diff
@@ -4,135 +4,135 @@ Extensions
 ----------
 
 Fundamentally, JupyterLab is designed as an extensible environment. JupyterLab
-extensions can customize or enhance any part of JupyterLab. They can provide new
-themes, file viewers and editors, or renderers for rich outputs in notebooks.
-Extensions can add items to the menu or command palette, keyboard shortcuts, or
-settings in the settings system. Extensions can provide an API for other
-extensions to use and can depend on other extensions. In fact, the whole of
-JupyterLab itself is simply a collection of extensions that are no more powerful
-or privileged than any custom extension.
+extensions can customize or enhance any part of JupyterLab. They can provide
+new themes, file viewers and editors, or renderers for rich outputs in
+notebooks. Extensions can add items to the menu or command palette, keyboard
+shortcuts, or settings in the settings system. Extensions can provide an API
+for other extensions to use and can depend on other extensions. In fact, the
+whole of JupyterLab itself is simply a collection of extensions that are no
+more powerful or privileged than any custom extension.
+
+We encourage extension authors to add the `jupyterlab-extension GitHub topic
+<https://github.com/search?utf8=%E2%9C%93&q=topic%3Ajupyterlab-extension&type=Repositories>`__
+to any GitHub extension repository.
+
+For information about developing extensions, see the :ref:`developer
+documentation <developer_extensions>`.
+
 
 .. contents:: Table of contents
     :local:
     :depth: 1
 
-Starting in JupyterLab 3.0, extensions are typically ``pip`` or ``conda``
-packages that are dynamically loaded by the application (no more build step). You can search using ``pip search "jupyterlab extension"`` to find extensions. For information about developing extensions,
-see the :ref:`developer documentation <developer_extensions>`.
-Once an extension is installed, you can refresh a running application
-or launch a new application to use the new extension.
+Installing Extensions
+~~~~~~~~~~~~~~~~~~~~~
 
-.. note::
+A JupyterLab extension is a JavaScript package that runs in the browser. An
+extension's JavaScript can be distributed as a single JavaScript package (which
+requires a rebuild of JupyterLab to activate it), or starting in JupyterLab
+3.0, can be built into a federated bundle together with its dependencies
```

Comment:
Add a definition of federated or a link to a definition? I've been following the update to jlab 3.0 and I'm still not sure I really know what federated means

Reply:
In an earlier draft, I even took out the word "federated" - we're making up that term to try to describe extensions that have been bundled together with their dependencies and don't require a lab rebuild, but perhaps we can just say that rather than invent a new kind of extension in people's mind.

One of the things that writing this down clarified for me was:

A. We have three different *types* of extensions - the normal generic extension, and two simpler ones that do a lot of the boilerplate for you (mime extension, theme extension)
B. An extension can come as a single npm package, or can be distributed as a bundle that includes its dependencies - this is what determines if lab needs to be rebuilt or not.
C. An extension can be packaged up and distributed via several different package managers: npm, pip, conda.

All three of these concepts are really independent of each other, so I think it's probably wise to not conflate them in our writeup.

Reply:
In an earlier draft, I even took out the word "federated" - we're making up that term to try to describe extensions that have been bundled together with their dependencies and don't require a lab rebuild, but perhaps we can just say that rather than invent a new kind of extension in people's mind.

One of the things that writing this down clarified for me was:

A. we have three different *types* of extensions - the normal generic extension, and two simpler ones that cut out a lot of the boilerplate: mime extensions and theme extensions.
B. Any extension may be distributed as a single js package (requiring a jlab build), or may be bundled together with its dependencies (not requiring a jlab build), or both.
C. Any extension can be packaged in the npm, conda, or pip package managers.

These three concepts are really independent of each other, and I think it does well for us to not conflate them. When we've used the term "federated extension" in the past, we've been referring to choices in both B (bundled together with its dependencies, not needing a jlab rebuild) and C (distributed as a pip or conda package).

I was debating even using the word "federated" in this writeup, but in the end did because I think we use the term in our codebase. I'm not sure we should use the term, or if we do, we make sure that we only use it in the context of B.

Reply:
I pushed a change that eliminates the word "federated" to try for more simplicity. Thoughts?

Reply:
I guess the term `federated` has been pull from the webpack5 federation terminology. It it the perfect time to converge to an easy terminology. Let's say I create from the cookie cutter an simple extension. We have the following scenarii:

- `jupyter labextension link .` and develop with wath
- `jupyter labextension .`
- `yarn watch`
- Publish in pip and `pip install` it (does not need a jupyter lab build)
- Publish in npm and `jupyter labextension install` it (needs a jupyter lab build with node.js)

It is exactly the same code, but used in various ways. Do we need separated names for them?

Reply:
> I guess the term `federated` has been pulled from the webpack5 federation terminology.

Yes, that's where we got the term.



> Do we need separated names for them?

I think it makes sense to have ways of talking about the separate concerns of A, B, and C from above. A determines what the extension author writes. B determines if a rebuild is needed. C determines how the end-user gets the package.

> `jupyter labextension link .` and develop with watch

B - single package, and C - distributed as a file, not via npm/pip/conda

> jupyter labextension install .

B - single package, and C - distributed as a file, not via npm/pip/conda

> Publish in pip and pip install it (does not need a jupyter lab build)

B - bundle, and C - distributed via pip

> Publish in pip and pip install it (does need a jupyter lab build)

B - single package, C - distributed via pip

> Publish in npm and jupyter labextension install it (needs a jupyter lab build with node.js)

B - single package, C - distributed via npm

> (hypothetical future) Publish in npm and jupyter labextension install it (does not need rebuild)

B - bundle (in, for example, a dist/ directory), C - distributed via npm

> (hypothetical future) Publish in npm and dynamically loaded from cdn like jsdelivr

B - bundle (in, for example, a dist/ directory), C - published to npm, obtained via cdn

Reply:
Maybe a table would help? (WIP, not taking all valid options listed above)

|                      | local     | npm     | pip     |
| -------------------- |:---------:|:-------:|:-------:|
| labextension link    | static    | -       | -       |
| yarn watch           | dynamic   | -       | -       | 
| labextension install | static    | static  | -       | 
| pip install          | dynamic  | static   | dynamic |

Reply:
I feel `federated` term should be replaced others, the above WIP table tells me something like `static` and `dynamic`. `static` means you need to rebuild a bit like the static linux libraries.

Reply:
You can have an npm extension that doesn't require rebuilding (though we don't support that quite yet, maybe in 3.1?), and you already have lots of pypi/conda extensions that do require a rebuild. You can develop an extension that bundles its dependencies or requires a rebuild (so the local install options could be either "static" or "dynamic").

Reply:
> You can have an npm extension that doesn't require rebuilding (though we don't support that quite yet, maybe in 3.1?),

I have updated the table with the above.


|                      | local     | npm     | pypi/conda |
| -------------------- |:---------:|:-------:|:----------:|
| labextension link    | static    | -       | -          |
| yarn watch           | dynamic   | -       | -          | 
| labextension install | static    | static<br/> (dynamic in 3.1)  | -          | 
| pip install          | dynamic   | -       | dynamic    | 

> you already have lots of pypi/conda extensions that do require a rebuild

Are you referring to the server side of existing jupyterlab2 extensions. I was reading that table in the jupyterlab3 world, not in jupyterlab2 See also https://github.com/jupyterlab/extension-cookiecutter-ts/issues/106 to would bring more clarity on that.

> You can develop an extension that bundles its dependencies or requires a rebuild (so the local install options could be either "static" or "dynamic").

Are you referring to extensions that bundle themselve with webpack? How would it impact the table?

Reply:
> Are you referring to the server side of existing jupyterlab2 extensions. I was reading that table in the jupyterlab3 world, not in jupyterlab2 See also [jupyterlab/extension-cookiecutter-ts#106](https://github.com/jupyterlab/extension-cookiecutter-ts/issues/106) to would bring more clarity on that.

Both are possible in JupyterLab 3. Again, static/dynamic is independent of how the extension is packaged/distributed. We shouldn't conflate the two. The cookie cutter picks only one possibility (python package, dynamic extension) as a recommended option.

Reply:
> Both are possible in JupyterLab 3. Again, static/dynamic is independent of how the extension is packaged/distributed. We shouldn't conflate the two. The cookie cutter picks only one possibility (python package, dynamic extension) as a recommended option.

A bit lost here. I understand now there are more options than I was surfacing, but not sure which command or configuration to invoke to concretize them. Maybe you could update the table if it makes sense?

Reply:
That's the thing. I don't think a table like that captures the independence of concerns here.

To choose between dynamic or static, the extension author runs the `jupyter labextension build` command in their extension (to generate the wepback federation bundle) or not. It has really nothing to do with which package manager is used (or if none at all is used).

Thanks, this conversation is helping things be clearer in my head about how we should explain the system so it is understandable (we're not there yet, but I think we're making progress).

Reply:
FWIW

|                      | local     | npm     | pypi/conda |
| -------------------- |:---------:|:-------:|:----------:|
| labextension link    | static    | -       | -          |
| yarn watch           | dynamic   | -       | -          | 
| labextension install | static    | static<br/> (dynamic in 3.1)  | -          | 
| pip install          | dynamic<br/>static   | -       | dynamic<br/>static    | 

(*) `static` implies `jupyter lab build` to be run to create the static bundle with all installed extensions.

Reply:
@jasongrout Is there currently any meaningful distinction between the `labextension install` and the `labextension link` commands? If not, can we finally drop `link`? Regardless, wrt this table for the docs I think it would be clearer if we drop `link` from the table/discussion for now.

Reply:
@telamonian - I think we still need `link` for packages that are not extensions. For packages that *are* extensions, link just calls install, if I'm reading things correctly: https://github.com/jupyterlab/jupyterlab/blob/f7cb7b5e208d732c8d841ae84d219cbebc97b22a/jupyterlab/commands.py#L938-L940

Reply:
I don't think we have a "dynamic" version of linked, i.e., federated bundles that are not extensions that are nevertheless loaded at runtime. There's no technical reason we can't, just IIRC it's not been implemented yet.

Reply:
@echarles - if I were to make a table to help clarify things, I think it would be something like:

| | local | npm | pypi/conda |
|-|-----|------|-------------|
| static | `jupyter labextension install ./local/path` | `jupyter labextension install npm-package` | `pip/conda install` |
| dynamic | See [here](https://github.com/jupyterlab/extension-cookiecutter-ts/tree/3.0/%7B%7Bcookiecutter.python_name%7D%7D#development-install), namely `jupyter labextension develop .` | not available yet | `pip/conda install` |

Reply:
> I feel federated term should be replace, the above WIP table tells me something like static and dynamic. static means you need to rebuild a bit like the static linux libraries.

Other terms that have been used until now are `bundled` / `federated` as an alternative to `static` / `dynamic`.

Also for reference there was a switch from `dynamic` to `federated` in this PR: https://github.com/jupyterlab/jupyterlab/pull/9050, which could be because "dynamic" sounded too generic?

Reply:
For context, @afshin didn't really like "dynamic", and we agreed to use "federated". It seems "federated" is also confusing to some.

As for "bundled" - that could mean "extensions which will be bundled into JupyterLab" or it could mean "extensions that come bundled with their dependencies".

Reply:
I do think this PR is big enough as it is, though. I'd suggest that if it seems to improve things, we merge and iterate as we continue this discussion of what terms to use.